backend/shade/shade.go

```diff
@@ -0,0 +1,1086 @@
+		return errors.New("cannot delete root directory")
+	}
+
+	fs.Debugf(f, "Encoded path for rmdir: %s", fullPath)
+	var response []api.ListDirResponse
+	res, err := f.callAPI(ctx, "GET", fmt.Sprintf("/%s/fs/listdir?path=%s", f.drive, fullPath), &response)
+
+	if res != nil && res.StatusCode != http.StatusOK {
+		fs.Debugf(f, "Failed to list dirs: status code %d", res.StatusCode)
+		return err
+	}
+
+	if len(response) > 0 {
+		return fs.ErrorDirectoryNotEmpty
+	}
+
+	// Use the delete endpoint which handles both files and directories
+	res, err = f.callAPI(ctx, "POST", fmt.Sprintf("/%s/fs/delete?path=%s", f.drive, fullPath), nil)
+	if err != nil {
+		fs.Debugf(f, "Error removing directory: %v", err)
+		return err
+	}
+	defer fs.CheckClose(res.Body, &err)
+
+	if res.StatusCode == http.StatusNotFound {
+		return fs.ErrorDirNotFound
+	}
+
+	if res.StatusCode != http.StatusOK && res.StatusCode != http.StatusCreated {
+		return fmt.Errorf("rmdir failed with status code: %d", res.StatusCode)
+	}
+
+	f.createdDirMu.Lock()
+	defer f.createdDirMu.Unlock()
+	unescapedPath, err := url.QueryUnescape(fullPath)
+	if err != nil {
+		return err
+	}
+	f.createdDirs[unescapedPath] = false
+
+	fs.Debugf(f, "Successfully removed directory: %s", dir)
+	return nil
+}
+
+// Attempts to construct the full path for an object query-escaped
+func (f *Fs) buildFullPath(remote string) string {
+	if f.root == "" {
+		return url.QueryEscape(remote)
+	}
+	return url.QueryEscape(path.Join(f.root, remote))
+}
+
+// -------------------------------------------------
+// Object implementation
+// -------------------------------------------------
+
+// Fs returns the parent Fs
+func (o *Object) Fs() fs.Info {
+	return o.fs
+}
+
+// String returns a description of the Object
+func (o *Object) String() string {
+	if o == nil {
+		return "<nil>"
+	}
+	return o.remote
+}
+
+// Remote returns the remote path
+func (o *Object) Remote() string {
+	return o.remote
+}
+
+// Hash returns the requested hash of the object content
+func (o *Object) Hash(ctx context.Context, t hash.Type) (string, error) {
+	if t != hash.MD5 {
+		return "", hash.ErrUnsupported
+	}
+	if o.Size() == 0 {
+		return "d41d8cd98f00b204e9800998ecf8427e", nil
+	}
+
+	var body io.ReadCloser
+	var err error
+
+	if o.original != "" {
+		resp, err := http.Get(o.original)
+		if err != nil {
+			return "", err
+		}
+		defer func() {
+			if cerr := resp.Body.Close(); cerr != nil && err == nil {
+				err = fmt.Errorf("failed to close response body: %w", cerr)
+			}
+		}()
+		if resp.StatusCode != http.StatusOK {
+			return "", fmt.Errorf("unexpected status %d fetching %s", resp.StatusCode, o.original)
+		}
+		body = resp.Body
+	} else {
+		body, err = o.Open(ctx)
+		if err != nil {
+			return "", err
+		}
+		defer func() {
+			if cerr := body.Close(); cerr != nil && err == nil {
+				err = fmt.Errorf("failed to close body: %w", cerr)
+			}
+		}()
+	}
+
+	h := md5.New()
+	if _, err := io.Copy(h, body); err != nil {
```

Comment:
Don't read the file to calculate the hash - that is the job of other parts of rclone. Just `return "", hash.ErrUnsupported`